File: api/services/byoapi.service.ts
import axios, { AxiosResponse } from 'axios';

import { ByoApiConfig } from '../models/byoApi.config.model';
import { CreatureModel } from '../models/creature.model';
import { CreatureViewModel } from '../view-models/creature.view-model';

const config = require('config');

export class ByoApiService {

    public static async searchForCreatures(queryString: string) : Promise<CreatureViewModel[]> {
        let creatures: CreatureViewModel[] = [];
        let endpoints: ByoApiConfig[] = config.get('byoapis');

        let calls: Record<string, Promise<AxiosResponse<CreatureModel[], any>>> = {};
        endpoints.forEach(endpoint => {
            calls[endpoint.id] = axios.get(`${endpoint.baseUrl}creatures?${queryString}`);
        });

        var callsToAwait = Object.entries(calls).map(([k,v]) => v);
        await Promise.all(callsToAwait);

        Object.entries(calls).map(([k,v]) => k).forEach(async (key) => {
            let data = (await calls[key]).data;
            data.forEach(x => { 
                let creature: CreatureViewModel = { byoapiId: key, ...x };
                creatures.push(creature); 
            });
        });

        return creatures;
    }

    public static async getCreatureById(id: string, byoapiId: string) : Promise<CreatureViewModel> {
        let creature: CreatureViewModel;
        let endpoint: ByoApiConfig = config.get('byoapis').find((x: ByoApiConfig) => x.id === byoapiId);
        let result = await axios.get(`${endpoint.baseUrl}creatures/${id}`);

        creature = result.data as CreatureViewModel;

        return creature;
    }
}

File: backends/DDD.3ncount3r.API/Services/ByoapiService.cs
using DDD._3ncount3r.API.Configurations;
using DDD._3ncount3r.API.Models;
using Microsoft.Extensions.Options;

namespace DDD._3ncount3r.API.Services
{
  public class ByoapiService: IByoapiService
  {
    private readonly IEnumerable<ByoapiConfig> _config;

    public ByoapiService(IOptions<List<ByoapiConfig>> options)
    {
      _config = options.Value;
    }

    public async Task<IEnumerable<CreatureModel>> SearchForCreatures(string queryString)
    {
      List<CreatureModel> allCreatures = new List<CreatureModel>();

      Dictionary<string, Task<HttpResponseMessage>> calls = new Dictionary<string, Task<HttpResponseMessage>>();

      HttpClient client = new HttpClient();
      foreach (var byoapi in _config)
      {        
        calls.Add(byoapi.Id, client.GetAsync($"{byoapi.BaseUrl}creatures{queryString}"));
      }

      await Task.WhenAll(calls.Values);

      foreach (var call in calls)
      {
        if (call.Value.IsCompletedSuccessfully)
        {
          HttpResponseMessage response = call.Value.Result;

          if (response.IsSuccessStatusCode)
          {
            allCreatures.AddRange(await response.Content.ReadFromJsonAsync<List<CreatureModel>>());
            allCreatures.ForEach(creature => { creature.ByoapiId = call.Key; });
          }
        }
      }

      return allCreatures;
    }

    public async Task<CreatureModel> GetCreatureByName(string byoapiId, string name)
    {
      ByoapiConfig endpoint = _config.FirstOrDefault(x => x.Id == byoapiId);

      if (endpoint == null) return new CreatureModel();

      HttpClient client = new HttpClient();
      HttpResponseMessage response = await client.GetAsync($"{endpoint.BaseUrl}creatures/{name}");

      response.EnsureSuccessStatusCode();

      return await response.Content.ReadFromJsonAsync<CreatureModel>();
    }
  }
}
